src/handshakes/ik.ts

```diff
@@ -0,0 +1,172 @@
+import {Buffer} from "buffer";
+import {x25519} from "bcrypto";
+
+import {CipherState, HandshakeState, MessageBuffer, NoiseSession, SymmetricState} from "../@types/handshake";
+import {bytes, bytes32} from "../@types/basic";
+import {generateKeypair, getHkdf} from "../utils";
+import {AbstractHandshake} from "./abstract-handshake";
+import {KeyPair} from "../@types/libp2p";
+import {BN} from "bn.js";
+
+
+export class IKHandshake extends AbstractHandshake {
+  public initSession(initiator: boolean, prologue: bytes32, s: KeyPair, rs: bytes32): NoiseSession {
+    const psk = this.createEmptyKey();
+
+    let hs;
+    if (initiator) {
+      hs = this.initializeInitiator(prologue, s, rs, psk);
+    } else {
+      hs = this.initializeResponder(prologue, s, rs, psk);
+    }
+
+    return {
+      hs,
+      i: initiator,
+      mc: new BN(0),
+    };
+  }
+
+  public sendMessage(session: NoiseSession, message: bytes): MessageBuffer {
+    let messageBuffer: MessageBuffer;
+    if (session.mc.eqn(0)) {
+      messageBuffer = this.writeMessageA(session.hs, message);
+    } else if (session.mc.eqn(1)) {
+      const { messageBuffer: mb, h, cs1, cs2 } = this.writeMessageB(session.hs, message);
+      messageBuffer = mb;
+      session.h = h;
+      session.cs1 = cs1;
+      session.cs2 = cs2;
+    } else if (session.mc.gtn(1)) {
+      if (session.i) {
+        if (!session.cs1) {
+          throw new Error("CS1 (cipher state) is not defined")
+        }
+
+        messageBuffer = this.writeMessageRegular(session.cs1, message);
+      } else {
+        if (!session.cs2) {
+          throw new Error("CS2 (cipher state) is not defined")
+        }
+
+        messageBuffer = this.writeMessageRegular(session.cs2, message);
+      }
+    } else {
+      throw new Error("Session invalid.")
+    }
+
+    session.mc = session.mc.add(new BN(1));
+    return messageBuffer;
+  }
+
+  public recvMessage(session: NoiseSession, message: MessageBuffer): bytes {
+    let plaintext: bytes;
+    if (session.mc.eqn(0)) {
+      plaintext = this.readMessageA(session.hs, message);
+    } else if (session.mc.eqn(1)) {
+      const { plaintext: pt, h, cs1, cs2 } = this.readMessageB(session.hs, message);
+      plaintext = pt;
+      session.h = h;
+      session.cs1 = cs1;
+      session.cs2 = cs2;
```

Comment:
Aren't you suppose to clear session.hs here?

Reply:
I am not sure if I am supposed to but I guess I could as it is not required anymore.